### PR TITLE
Autocomplete - Phase 1

### PR DESCRIPTION
**Pull request by danielstro**
Adding <Autocomplete /> . Current phase only supports filtering results and onClick / onChange

### CODE REVIEW

File: src/components/auto-complete/auto-complete.tsx [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    }
+    return null;
+}, style);
+
+export interface AutoCompleteProps extends OptionList {
+    value?: string;
+    onChange?: (value: string) => void;
+    style?: any;
+    className?: string;
+    onItemClick?: (item: string) => void;
```
Comment (tobich) [inline src/components/auto-complete/auto-complete.tsx]: I don't think this is a part of the spec

File: src/components/auto-complete/auto-complete.tsx [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+                    dataSource={props.items}
+                    onChange={props.onItemClick!}
+                />
+            </div>
+        );
+    }
+    return null;
+}, style);
+
+export interface AutoCompleteProps extends OptionList {
```
Comment (tobich) [inline src/components/auto-complete/auto-complete.tsx]: Why not use `Partial<>` instead of all the question marks?

File: src/components/auto-complete/auto-complete.tsx [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+    value?: string;
+    onChange?: (value: string) => void;
+    style?: any;
```
Comment (tobich) [inline src/components/auto-complete/auto-complete.tsx]: `style` and `className` shouldn't be explicit props. We'll deal with them on a general level.

File: src/components/auto-complete/auto-complete.st.css [hunk 4]
```diff
@@ -0,0 +1,20 @@
+.root {
```
Comment (tobich) [inline src/components/auto-complete/auto-complete.st.css]: Please make sure the font in the input field looks the same as in the list.

File: src/components/auto-complete/auto-complete.tsx (hunk 3 above)
Comment (danielstro) [inline src/components/auto-complete/auto-complete.tsx]: removed them for now

File: src/components/auto-complete/auto-complete.tsx (hunk 1 above)
Comment (danielstro) [inline src/components/auto-complete/auto-complete.tsx]: yeah, I changed it to use onChange instead

File: src/components/auto-complete/auto-complete.st.css (hunk 4 above)
Comment (danielstro) [inline src/components/auto-complete/auto-complete.st.css]: done

File: src/components/auto-complete/auto-complete.tsx (hunk 2 above)
Comment (danielstro) [inline src/components/auto-complete/auto-complete.tsx]: Changed it to partial but the ? were like that before so I think that if we want to change them we should do it in a different PR

File: demo/components/auto-complete.demo.tsx [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    inputText: string;
+}
+
+const items = ['Muffins', 'Pancakes', 'Cupcakes', 'Souffles',
+               'Pasta', 'Soup', 'Caramel', 'Avazim', 'Moses'];
+
+export class AutoCompleteDemo extends React.Component<{}, AutoCompleteDemoState> {
+    public state = {open: true, inputText: ''};
+
+    public onChange = (e: any) => {
```
Comment (liorzisman) [inline demo/components/auto-complete.demo.tsx]: type

File: src/components/auto-complete/auto-complete.tsx [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    items?: string[];
+    onChange?: (item: string) => void;
+}
+
+export interface AutoCompleteChangeEvent extends React.ChangeEvent<HTMLInputElement> {
+    value: string;
+}
+
+export const AutoCompleteList: React.SFC<AutoCompleteListProps> = SBStateless(props => {
+    if (props.open) {
```
Comment (liorzisman) [inline src/components/auto-complete/auto-complete.tsx]: probably more fluent to have `if (!props.open) return null;` and then the rest of the return

File: src/components/auto-complete/auto-complete.tsx [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+
+export interface AutoCompleteProps extends React.InputHTMLAttributes<HTMLInputElement>, Partial<OptionList> {
+    open?: boolean;
+    value?: string;
+    onChange?: (event: Partial<AutoCompleteChangeEvent>) => void;
+    filter?: FilterPredicate;
+}
+
+export interface AutoCompleteState {
+    isOpen: boolean;
```
Comment (liorzisman) [inline src/components/auto-complete/auto-complete.tsx]: need to consider using this totally as a controlled prop

File: src/components/auto-complete/auto-complete.tsx [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+        filter: prefixFilter,
+        onChange: noop
+    };
+    public state = {input: null, isOpen: this.props.open!};
+
+    public render() {
+        const rootProps = root(this.props, {
+            'data-automation-id': 'AUTO_COMPLETE',
+            'className': 'auto-complete'
+        }) as React.HtmlHTMLAttributes<HTMLDivElement>;
```
Comment (liorzisman) [inline src/components/auto-complete/auto-complete.tsx]: `HtmlHTMLAttributes`?

Comment (liorzisman) [inline src/components/auto-complete/auto-complete.tsx]: `HtmlHTMLAttributes`?

File: src/components/auto-complete/auto-complete.tsx [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            <SelectionList
+                className="auto-complete-list"
+                dataSource={props.items}
+                onChange={props.onChange!}
+            />
+        </div>
+    );
+}, style);
+
+export interface AutoCompleteProps extends React.InputHTMLAttributes<HTMLInputElement>, Partial<OptionList> {
```
Comment (JoMarton) [inline src/components/auto-complete/auto-complete.tsx]: at this moment we don't implement `React.InputHTMLAttributes<HTMLInputElement>` interface. use `FormInputProps<string>` interface imported from `src/types/forms.ts`.
This will already include definition for `value?: string` and `onChange?: ({value: string}) => void`

File: src/components/auto-complete/auto-complete.st.css [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    width: 100%;
+}
+
+.caret {
+    cursor: pointer;
+    position:absolute;
+    top:50%;
+    left: 90%;
+    transform:translateY(-50%);
+    -webkit-transform:translateY(-50%);
```
Comment (tobich) [inline src/components/auto-complete/auto-complete.st.css]: Why -webkit- prefix?

File: src/components/auto-complete/auto-complete.st.css [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    padding-left: 5px;
+}
+
+.auto-complete-list {
+    -st-extends: List;
+    background: #FFFFFF;
+}
+
+.auto-complete-input:hover {
+    cursor: pointer;
```
Comment (tobich) [inline src/components/auto-complete/auto-complete.st.css]: Please consult @giladgrush if there should be really a "pointer" over the input. methinks it should be "caret".

File: demo/components/auto-complete.demo.tsx [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+export interface AutoCompleteDemoState {
+    open: boolean;
+    inputText: string;
+}
+
+const items = ['Muffins', 'Pancakes', 'Cupcakes', 'Souffles',
+               'Pasta', 'Soup', 'Caramel', 'Avazim', 'Moses'];
+
+export class AutoCompleteDemo extends React.Component<{}, AutoCompleteDemoState> {
+    public state = {open: true, inputText: ''};
```
Comment (tobich) [inline demo/components/auto-complete.demo.tsx]: Why does the control in the demo start with an open popup? It looks strange.

Comment (danielstro) [reply demo/components/auto-complete.demo.tsx]: It's a leftover from when the caret wasn't there. Fixed it

File: src/components/auto-complete/auto-complete.st.css (hunk 10 above)
Comment (danielstro) [reply src/components/auto-complete/auto-complete.st.css]: for mobile

File: test/components/auto-complete.spec.tsx [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+        it('types in the input and selects a value', async () => {
+            const {select, waitForDom} = clientRenderer.render(<AutoCompleteDemo />);
+            const itemList = bodySelect(list)!;
+            const prefix = 'P';
+            const filteredItems = items.filter(item => item.startsWith(prefix)).join('');
+
+            await waitFor(() => expect(itemList.textContent).to.eql(items.join('')));
+
+            bodySelect<HTMLInputElement>(autoCompInput)!.value = prefix;
+            simulate.change(bodySelect(autoCompInput));
```
[the remark this answers is not in the document]
Comment (danielstro) [reply test/components/auto-complete.spec.tsx]: done

File: test/components/auto-complete.spec.tsx [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+            await waitFor(() => expect(itemList.textContent).to.eql(filteredItems));
```
[the remark this answers is not in the document]
Comment (danielstro) [reply test/components/auto-complete.spec.tsx]: done

File: test/components/auto-complete.spec.tsx [hunk 15]
```diff
@@ -0,0 +1,113 @@
...
+        const {select, waitForDom} = clientRenderer.render(<AutoComplete />);
+
+        return waitForDom(() => expect(select(autoComp)).to.be.present());
+    });
+
+    it('invokes the onChange when text is entered to label', () => {
+        const onChange = sinon.spy();
+        const {select} = clientRenderer.render(<AutoComplete onChange={onChange}/>);
+
+        (select(autoComp, autoCompInput) as HTMLInputElement).value = 'abc';
```
[the remark this answers is not in the document]
Comment (danielstro) [reply test/components/auto-complete.spec.tsx]: done

File: test/components/auto-complete.spec.tsx [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+            await waitForDom(() => {
+                expect(select(autoCompDemo + '_TEXT')).to.have.text('You picked: Pancakes');
+            });
+        });
+    });
+
+    it('renders to the screen', () => {
+        const {select, waitForDom} = clientRenderer.render(<AutoComplete />);
+
+        return waitForDom(() => expect(select(autoComp)).to.be.present());
```
[the remark this answers is not in the document]
Comment (danielstro) [reply test/components/auto-complete.spec.tsx]: done

File: test/components/auto-complete.spec.tsx [hunk 17]
```diff
@@ -0,0 +1,113 @@
...
+    it('renders the item list if open is given', async () => {
+        clientRenderer.render(<AutoComplete open/>);
+
+        await waitFor(() => expect(bodySelect(list)).to.be.present());
+    });
+
+    it('renders the items if given', async () => {
+        clientRenderer.render(<AutoComplete open dataSource={items}/>);
+
+        await waitFor(() => expect(bodySelect(list, 'LIST')!.children[0].innerHTML).to.be.equal('Muffins'));
```
[the remark this answers is not in the document]
Comment (danielstro) [reply test/components/auto-complete.spec.tsx]: It will, but only partially.

File: test/components/auto-complete.spec.tsx [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+        });
+    });
+
+    it('displays filtered results according to given value', async () => {
+        const prefix = 'P';
+        clientRenderer.render(<AutoComplete open dataSource={items} value={prefix}/>);
+        const itemList = bodySelect(list);
+
+        await waitFor(() => {
+            expect(itemList!.textContent).to.eql(items.filter(item => item.startsWith(prefix)).join(''));
```
[the remark this answers is not in the document]
Comment (danielstro) [reply test/components/auto-complete.spec.tsx]: done

File: test/components/auto-complete.spec.tsx [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+    afterEach(() => clientRenderer.cleanup());
+
+    describe('The autocomplete user', () => {
+        it('types in the input and selects a value', async () => {
+            const {select, waitForDom} = clientRenderer.render(<AutoCompleteDemo />);
+            const itemList = bodySelect(list)!;
+            const prefix = 'P';
+            const filteredItems = items.filter(item => item.startsWith(prefix)).join('');
+
+            await waitFor(() => expect(itemList.textContent).to.eql(items.join('')));
```
[the remark this answers is not in the document]
Comment (danielstro) [reply test/components/auto-complete.spec.tsx]: done

File: src/components/auto-complete/auto-complete.st.css (hunk 11 above)
Comment (danielstro) [reply src/components/auto-complete/auto-complete.st.css]: it shouldn't. this doesn't belong here

File: src/components/auto-complete/auto-complete.st.css (hunk 10 above)
Comment (tobich) [reply src/components/auto-complete/auto-complete.st.css]: which mobile requires -webkit-? Is it failing in the ci?

File: demo/components-demo.tsx [hunk 20]
```diff
@@ -1,4 +1,5 @@
 import * as React from 'react';
```
Comment (tobich) [inline demo/components-demo.tsx]: General remarks:
1. The "down" icon is misplaced
2. The behavior is weird. IMHO when I start writing, the dropdown should open automatically if there are hits

File: src/components/auto-complete/auto-complete.tsx (hunk 7 above)
Comment (tobich) [reply src/components/auto-complete/auto-complete.tsx]: This comment should be properly addressed. I also believe "isOpen" should be simple controlled property.

File: test/components/auto-complete.spec.tsx [hunk 21]
```diff
@@ -0,0 +1,120 @@
...
+    describe('The autocomplete user', () => {
+        it('types in the input and selects a value', async () => {
+            const {select, waitForDom} = clientRenderer.render(<AutoCompleteDemo />);
+            const prefix = 'P';
+            const filteredItems = items.filter(item => item.startsWith(prefix)).join('');
+
+            simulate.click(select(autoComp + '_CARET'));
+
+            const itemList = bodySelect(list)!;
+            await waitFor(() => {
```
Comment (tobich) [inline test/components/auto-complete.spec.tsx]: why waitFor and not waitForDom?

File: test/components/auto-complete.spec.tsx [hunk 22]
```diff
@@ -0,0 +1,120 @@
...
+                expect(itemList).to.be.present();
+                expect(itemList.textContent).to.equal(items.join(''));
+            });
+
+            trigger.change(bodySelect(autoCompInput), prefix);
+            await waitFor(() => expect(itemList.textContent).to.equal(filteredItems));
```
Comment (tobich) [inline test/components/auto-complete.spec.tsx]: waitForDom

File: test/components/auto-complete.spec.tsx [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+    it('renders the item list if open is given', async () => {
+        clientRenderer.render(<AutoComplete open/>);
+
+        await waitFor(() => expect(bodySelect(list)).to.be.present());
+    });
+
+    it('renders the items if given', async () => {
+        clientRenderer.render(<AutoComplete open dataSource={items}/>);
+
+        await waitFor(() => expect(bodySelect(list, 'LIST')!.children[0]).to.have.text('Muffins'));
```
Comment (tobich) [inline test/components/auto-complete.spec.tsx]: waitForDom

File: test/components/auto-complete.spec.tsx [hunk 24]
```diff
@@ -0,0 +1,120 @@
...
+
+    it('invokes the onChange when an option is clicked and closes the popup', async () => {
+        const onChange = sinon.spy();
+        clientRenderer.render(
+            <AutoComplete open dataSource={['Cat', 'Dog']} onChange={onChange}/>
+        );
+
+        simulate.click(bodySelect(list, 'LIST')!.children[0]);
+
+        await waitFor(() => {
```
Comment (tobich) [inline test/components/auto-complete.spec.tsx]: waitForDom

File: test/components/auto-complete.spec.tsx [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+            expect(bodySelect(list)).to.be.absent();
+        });
+    });
+
+    it('displays filtered results according to given value', async () => {
+        const prefix = 'P';
+        clientRenderer.render(<AutoComplete open dataSource={items} value={prefix}/>);
+        const itemList = bodySelect(list);
+
+        await waitFor(() => {
```
Comment (tobich) [inline test/components/auto-complete.spec.tsx]: waitForDom

File: test/components/auto-complete.spec.tsx [hunk 26]
```diff
@@ -0,0 +1,120 @@
...
+            expect([input, caret]).to.be.verticallyAligned('center');
+        });
+    });
+
+    it('opens the popup when clicking on the caret', async () => {
+        const {select, waitForDom} = clientRenderer.render(<AutoComplete/>);
+
+        await waitForDom(() => expect(select(autoComp, autoComp + '_CARET')).to.be.present());
+        simulate.click(select(autoComp, autoComp + '_CARET'));
+        await waitFor(() => expect(bodySelect(list)).to.be.present());
```
Comment (tobich) [inline test/components/auto-complete.spec.tsx]: waitForDom

File: src/components/auto-complete/auto-complete.st.css (hunk 10 above)
Comment (danielstro) [reply src/components/auto-complete/auto-complete.st.css]: Yes, we are running tests on android and it needs the webkit prefix to run

File: demo/components/auto-complete.demo.tsx [hunk 27]
```diff
@@ -0,0 +1,43 @@
...
+export class AutoCompleteDemo extends React.Component<{}, AutoCompleteDemoState> {
+    public state = {open: false, inputText: ''};
+
+    public onChange = (e: ChangeEvent<string>) => {
+        this.setState({
+            inputText: e.value
+        });
+    }
+
+    public toggleOpen = (e: ChangeEvent<boolean>) => {
```
Comment (tobich) [inline demo/components/auto-complete.demo.tsx]: This means we have implemented incorrectly controlled component. This event should just set the "open" state to `e.value`. The demo shouldn't do actual toggling! It should just pass the state. Something's wrong in the component implementation. We have btw the same mistake in the dropdown demo.
